python/ray/data/_internal/iter_batches.py

```diff
@@ -0,0 +1,934 @@
+            performance for non-CPU bound UDFs, allowing batch fetching compute and
+            formatting to be overlapped with the UDF. Defaults to 0 (no prefetching
+            enabled).
+
+    Returns:
+        An iterator over record batches.
+    """
+    context = DatasetContext.get_current()
+
+    if (
+        prefetch_batches > 0
+        and context.actor_prefetcher_enabled
+        and not ray.util.client.ray.is_connected()
+    ):
+        prefetcher = ActorBlockPrefetcher()
+    else:
+        prefetcher = WaitBlockPrefetcher()
+
+    def _async_iter_batches(block_refs):
+        # Step 1: Construct logical batches based on the metadata.
+        batch_iter = _bundle_block_refs_to_logical_batches(
+            block_refs, batch_size=batch_size, drop_last=drop_last
+        )
+
+        # Step 2: Shuffle the logical batches if applicable.
+        if shuffle_buffer_min_size is not None:
+            batch_iter = _local_shuffle_logical_batches(
+                shuffle_buffer_min_size=shuffle_buffer_min_size,
+                shuffle_seed=shuffle_seed,
+            )
+
+        # Step 3: Prefetch logical batches locally.
+        if prefetch_batches > 0:
+            batch_iter = _prefetch_batches_locally(
+                batch_iter,
+                prefetcher=prefetcher,
+                num_batches_to_prefetch=prefetch_batches,
+                stats=stats,
+            )
+
+        def threadpool_computations(logical_batch_iter: Iterator[LogicalBatch]):
+            # Step 4.1: Resolve the blocks.
+            resolved_batch_iter = _resolve_blocks(
+                logical_batch_iter,
+                clear_block_after_read=clear_block_after_read,
+                stats=stats,
+            )
+
+            # Step 4.2: Slice the blocks to create the batch.
+            batch_iter = _construct_batch_from_logical_batch(
+                resolved_batch_iter, stats=stats, ensure_copy=ensure_copy
+            )
+
+            # Step 4.3: Format the batches.
+            formatted_batch_iter = _format_batches(
+                batch_iter, batch_format=batch_format, stats=stats
+            )
+
+            # Step 4.4: Apply the collate function if applicable.
+            if collate_fn is not None:
+                formatted_batch_iter = _collate(
+                    formatted_batch_iter, collate_fn=collate_fn, stats=stats
+                )
+            yield from formatted_batch_iter
+
+        # Step 4: Use a threadpool for resolving blocks, slicing, formatting, and
+        # collation.
+        if prefetch_batches > 0:
+            batch_iter = _make_async_gen(
+                batch_iter, fn=threadpool_computations, num_workers=prefetch_batches
+            )
+            # Step 5: Make sure to preserve order from threadpool results.
+            yield from _preserve_order(batch_iter)
+        else:
+            # If no batch prefetching is specified, then don't use a threadpool.
```

Comment:
It may be preferable to always use a threadpool, but use size 1 if there isn't any prefetching. This avoids two different code paths being in play.